### PR TITLE
Allow same-hostport-different-protocol

### LINKED ISSUE #8372: Duplicate hostPorts Rejected Despite Disparate Protocols
Pods which use `hostPort` to bind both TCP and UDP sockets to the same port number are rejected with the error:

```
ERROR: Error: Pod "foo" is invalid: spec.containers[0].port: duplicate value '2049'
```

An example pod follows.

```yaml
apiVersion: v1beta3
kind: Pod
metadata:
  name: foo
spec:
  containers:
    - name: foo
      image: foo/bar
      ports:
        - name: foo-tcp
          containerPort: 2049
          protocol: tcp
          hostPort: 2049
        - name: foo-udp
          containerPort: 2049
          protocol: udp
          hostPort: 2049
```

### PR DESCRIPTION
**Pull request by thockin**
Fixes #8372

### CODE REVIEW

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -673,22 +673,23 @@ func validateProbe(probe *api.Probe) errs.ValidationErrorList {
...
 		for pi := range ctr.Ports {
-			port := extract(&ctr.Ports[pi])
+			port := ctr.Ports[pi].HostPort
 			if port == 0 {
 				continue
 			}
-			if accumulator[port] {
-				cErrs = append(cErrs, errs.NewFieldDuplicate("port", port))
+			str := fmt.Sprintf("%d/%s", port, ctr.Ports[pi].Protocol)
+			if accumulator[str] {
```
Comment (bprashanth) [inline pkg/api/validation/validation.go]: suggest using util.StringSet and Has if you only want to check set membership

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -673,22 +673,23 @@ func validateProbe(probe *api.Probe) errs.ValidationErrorList {
 	return allErrs
 }
 
-// AccumulateUniquePorts runs an extraction function on each Port of each Container,
+// accumulateUniquePorts runs an extraction function on each Port of each Container,
```
Comment (bprashanth) [inline pkg/api/validation/validation.go]: the part about the extraction function isn't accurate anymore?

**Comment by bprashanth** [comment]
Shippable looks like it might be related:
--- FAIL: TestHandlePortConflicts (0.00s)
	kubelet_test.go:3015: status of pod "newpod_foo" is not found in the status map

**Comment by thockin** [comment]
New push is up

**Comment by thockin** [comment]
No, the accumulator (I assume you mean) has to outlive any one call so it can accumulate across pods.

**Comment by bprashanth** [comment]
Ah, blindness, LGTM will merge after rebase/green

**Comment by quinton-hoole** [comment]
Thanks for picking this up beeps.

**Comment by thockin** [comment]
Shippable is green

**Comment by bprashanth** [comment]
Waiting on https://github.com/GoogleCloudPlatform/kubernetes/issues/8534

**Pull request merged**